---
title: "Lieber 2024 TRACE ISA"
author: "Bob Policastro"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, collapse=TRUE, comment="#>")
```

## Libraries

Load the required libraries for the analysis.

```{r, message=FALSE, warning=FALSE}
library("tidyverse")
library("patchwork")
library("randomcoloR")
library("ggseqlogo")
library("Biostrings")
library("karyoploteR")
library("GenomicFeatures")
library("GenomicRanges")
library("GenomeInfoDb")
library("Rsamtools")
library("readr")
library("ChIPseeker")
```

## Prepare Data

Load the annotated insertion sites from the Ensoma TRACE ISA workflow.

```{r}
# Get a list of files to load.
isa_files <-
  file.path("data", "insertion_sites") |>
  list.files(pattern="filtered\\.tsv$", full.names=TRUE)

# Load the files.
isa <- isa_files |>
  tibble(file_path=_) |>
  mutate(
    file_name=basename(file_path),
    data=map(file_path, read_tsv, show_col_types=FALSE)
  ) |>
  unnest(data) |>
  # Remove the Illumina stuff from the sample names.
  mutate(file_id=str_remove(sample_id, "_S\\d+_L00\\d\\.\\S+$")) |>
  relocate(file_id, .after=sample_id)
```

Load the sample sheet that has additional information for each sample.
Then join the information with the ISA results.

```{r}
# Load the sample sheet.
sample_sheet <- read_csv(
  file.path("data", "sample_sheet.csv"), show_col_types=FALSE
)

# Join into the isa data.
isa <- left_join(isa, sample_sheet, by="file_id")
```

Only keep insertion sites that appear in at least 2 of 3 technical
replicates. Return the sum of values for the group of technical
replicates.

```{r}
isa_old <- isa

isa <- isa |>
  # Keep only insertion sites that appear in at least 2 technical replicates.
  group_by(sample_name, seqnames, start, end, strand) |>
  filter(n() > 1) |>
  ungroup() |>
  # Sum the score for each insertion site across replicates.
  group_by(
    sample_name, seqnames, start, end, strand, annotation, tss_distance 
  ) |>
  summarize(score=sum(score), .groups="drop")
```

Annotate the insertion sites.

```{r}
# Load the annotation as a TxDb.
txdb <- 
  file.path("data", "Mus_musculus.GRCm39.111.chr.gtf.gz") |>
  makeTxDbFromGFF()

# Reannotate the insertion sites.
isa <- isa |>
  dplyr::select(!c(annotation, tss_distance)) |>
  makeGRangesFromDataFrame(keep.extra.columns=TRUE) |>
  annotatePeak(
    peak=_,
    tssRegion=c(-2000, 200),
    TxDb=txdb,
    level="transcript",
    sameStrand=FALSE
  ) |>
  as_tibble() |>
  dplyr::select(
    sample_name, seqnames, start, end, strand, annotation,
    tss_distance=distanceToTSS, score
  )
```

Export the filtered insertion sites for inclusion in the final paper.

```{r}
isa_old |> 
  dplyr::select(!c(
    file_path, file_name, sample_id, file_id, score, sample_name, annotation,
    tss_distance, score, sample_source, gene_marking_percent,
    integration_vector, payload_vector, species
  )) |>
  distinct() |>
  right_join(isa, by=c("seqnames", "start", "end", "strand"), multiple="all") |>
  relocate(sample_name) |>
  relocate(score, annotation, tss_distance, .after=strand) |>
  write_tsv(file.path("results", "TRACE_insertion_sites_annotated.tsv"))
```

Set the factor levels of the sample names so they appear in a consistent
order in the plots.

```{r}
sample_name_order <- c(
  "Mouse BM (65.10%)", "Mouse BM (62.20%)", "Mouse BM (38.90%)",
  "Mouse BM (37.00%)", "Mouse BM (25.00%)", "Mouse BM (0.00%)"
)

isa <- mutate(
  isa, sample_name=factor(sample_name, levels=sample_name_order)
)
```

## Number and Frequency of Insertion Sites

Generate a bar graph of the number of insertion sites per sample. This
is defined as the number of unique genomic locations for each sample
where an insertion event was detected.

```{r, fig.height=3.5, fig.width=4.25}
p1 <- isa |>
  count(sample_name) |>
  ggplot(aes(x=n, y=fct_rev(sample_name))) +
    geom_col(width=0.75, aes(fill=sample_name), show.legend=FALSE) +
    theme_bw() +
    labs(
      x="Unique Insertion Site\nLocations",
      y="Tissue (GFP+ %)",
      title="Unique Insertion Site\nLocations per Sample"
    ) +
    scale_fill_brewer(palette="Set2")

print(p1)

if (!dir.exists("results")) dir.create("results")

ggsave(
  file.path("results", "insertion_sites_per_sample.pdf"), plot=p1,
  device=cairo_pdf, height=3.5, width=4.25
)
```

Multiple unique sequenced fragments can end up mapping to the same
genomic location in the same sample. This is likely due to the clonal
expansion of a cell with a particular genomic insertion site. This
bargraph represents all fragments of DNA with insertion sites sequenced
for each sample regardless of genomic location.

```{r, fig.height=3.75, fig.width=4.25}
p2 <- isa |>
  # Sum the total number of unique fragments for each sample.
  group_by(sample_name) |>
  summarize(score=sum(score), .groups="drop") |>
  # Plot the total number of unique fragments for each sample.
  ggplot(aes(x=score, y=fct_rev(sample_name))) +
    geom_col(width=0.75, aes(fill=sample_name), show.legend=FALSE) +
    theme_bw() +
    theme(axis.text.x=element_text(angle=45, hjust=1)) +
    labs(
      x="Insertion Site-containing\nDNA Fragments Sequenced",
      y="Tissue (GFP+ %)",
      title="Insertion Site-containing\nDNA Fragments Sequenced\nper Sample"
    ) +
    scale_fill_brewer(palette="Set2")

print(p2)

ggsave(
  file.path("results", "insertion_sites_sequenced_per_sample.pdf"), plot=p2,
  device=cairo_pdf, height=3.5, width=4.25
)
```

Combine the previous two plots into one figure for publication.

```{r, fig.height=3.75, fig.width=7}
p <- 
  (p2 + p1) +
  plot_layout(axes="collect_y")

print(p)

ggsave(
  file.path("results", "insertion_sites_for_figure.pdf"), plot=p,
  device=cairo_pdf, height=3.75, width=7
)
```

As described above, a sample can have multiple DNA fragments with
insertion sites mapping to the same genomic location. This stacked
barplot represents for each genomic location with an insertion site in
the sample, the fraction of insertion-site containing DNA fragments
sequenced associated with that site.

```{r, fig.height=4.5, fig.width=5.5}
# Function to generate random color palette.
random_pal <- function(vals) {
  n_vals <- length(unique(vals))
  color_map <- unique(randomcoloR::randomColor(n_vals + 1000))[1:n_vals]
  names(color_map) <- unique(vals)
  colors <- vapply(vals, \(x) color_map[x], character(1))
  return(colors)
}

# Prepare data for plotting.
df <- isa |>
  # Summarize the fraction of reads per insertion site.
  dplyr::group_by(sample_name) |>
  dplyr::mutate(frac=score / sum(score)) |>
  dplyr::ungroup() |>
  # Add a column with the custom colors.
  dplyr::mutate(
      posid=paste(seqnames, start, end, strand, sep=":"),
      color=random_pal(posid),
      sample_id=factor(
          sample_name,
          levels=rev(stringr::str_sort(unique(sample_name)))
      )
  ) |>
  split(~sample_name) |>
  lapply(\(x) {
      x <- x |>
          dplyr::mutate(
              color=forcats::fct_drop(color),
              color=forcats::fct_reorder(color, frac, .desc=TRUE)
          )
      return(x)
  })

# Create the plots.
p <- df |>
  imap(\(x, y) {
    ggplot(x, aes(x=frac, y=sample_name)) +
      geom_col(aes(fill=color)) +
      scale_fill_identity() +
      scale_x_continuous(labels=scales::percent) +
      theme_classic() +
      labs(
        x="Percentage of Captured\nInsertion Events",
        y="Tissue (GFP+ %)"
      )
  }) |>
  wrap_plots(ncol=1) +
    plot_layout(axes="collect", axis_titles="collect")

print(p)

ggsave(
  file.path("results", "insertion_site_percentage_per_sample.pdf"),
  plot=p, device=cairo_pdf, height=4.5, width=5.5
)
```

## Genomic Distribution

As part of the Ensoma TRACE ISA workflow, each insertion site is
annotated to the nearest genomic feature at the *transcript* level.
Assignment to a feature is based on the the following rules:

-   **Promoter**: -2,000 to +200 bp relative to *transcript*
    transcription start sites (TSS).
-   **Exon**: Non-promoter and non-3\` UTR *exonic* positions.
-   **Intron**: Non-promoter and non-3\` UTR *intronic* positions.
-   **3\` UTR**: From the end of the ORF to the last exon of a
    transcript.
-   **Downstream**: +3,000 kbp from the end of the last exon of the
    downstream-most transcript exon.
-   **Distal Intergenic**: Not contained within the above constraints.

We first need to calculate the overall feature distribution of the mouse genome.
The first step is to download the assembly and get the total genome length.

```{r}
assembly_url <-
  "https://ftp.ensembl.org/pub/release-111/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz"

index_file <- file.path(
  "data", "Mus_musculus.GRCm39.dna.primary_assembly.fa.fai"
)

if (!file.exists(index_file)) {
  # Download the genome assembly.
    download.file(
      assembly_url,
      file.path("data", "Mus_musculus.GRCm39.dna.primary_assembly.fa.gz"),
      method="wget"
    )
  
  # Unzip the file.
  system2(
    "gunzip",
    file.path("data", "Mus_musculus.GRCm39.dna.primary_assembly.fa.gz")
  )
  
  # Index the genome assembly.
  indexFa(file.path("data", "Mus_musculus.GRCm39.dna.primary_assembly.fa"))
}
  
# Load the index.
genome_length <- index_file |>
  read_tsv(
    col_names=c(
      "seqnames", "width", "byte_index", "bases_per_line", "bytes_per_line"
    ),
    col_types=c("c", rep("i", 4))
  ) |>
  pull(width) |>
  sum()
```

Now that the total genome size has been obtained we can get the fraction of
the genome contained within each annotation. The priority of annotations is as
follows: Promoter, 3' UTR, exonic, intronic, downstream, and distal intergenic.

```{r}
# Download the mouse GTF file if it doesn't exist.
if (!file.exists(file.path("data", "Mus_musculus.GRCm39.111.chr.gtf.gz"))) {
  download.file(
    "https://ftp.ensembl.org/pub/release-111/gtf/mus_musculus/Mus_musculus.GRCm39.111.chr.gtf.gz",
    file.path("data", "Mus_musculus.GRCm39.111.chr.gtf.gz"),
    method="wget"
  )
}

# Load the GTF file as a TxDb object.
gtf <- makeTxDbFromGFF(file.path("data", "Mus_musculus.GRCm39.111.chr.gtf.gz"))

# Portion of genome that is promoter.
promoter_regions <- gtf |>
  transcripts() |>
  promoters(upstream=2000, downstream=200) |>
  GenomicRanges::reduce()

promoter_sum <- promoter_regions |>
  width() |>
  sum()

# Portions of genome that are 3' UTR but not promoter.
three_prime_utr_regions <- gtf |>
  threeUTRsByTranscript() |>
  unlist() |>
  GenomicRanges::reduce() |>
  GenomicRanges::setdiff(promoter_regions)

three_prime_utr_sums <- three_prime_utr_regions |>
  width() |>
  sum()

# Portion of genome that is exonic but not promoter or 3' UTR.
exonic_regions <- gtf |>
  exons() |>
  GenomicRanges::reduce() |>
  GenomicRanges::setdiff(c(promoter_regions, three_prime_utr_regions))

exonic_sum <- exonic_regions |>
  width() |>
  sum()

# Portion of the genome that is introniuc but not promoter, 3' UTR, or exonic.
intronic_regions <- gtf |>
  intronsByTranscript() |>
  unlist() |>
  GenomicRanges::reduce() |>
  GenomicRanges::setdiff(c(
    promoter_regions, three_prime_utr_regions, exonic_regions
  ))

intronic_sum <- intronic_regions |>
  width() |>
  sum()

# Portion of the genome that is downstream from the gene but not in any
# promoters, 3' UTRs, exonic, or intronic regions.
get_downstream_regions <- function(genes, distance) {
  downstream <- genes
  
  start(downstream)[as.character(strand(genes)) == "+"] <-
    end(genes)[as.character(strand(genes)) == "+"] + 1
  end(downstream)[as.character(strand(genes)) == "+"] <-
    end(genes)[as.character(strand(genes)) == "+"] + distance
  
  end(downstream)[as.character(strand(genes)) == "-"] <-
    start(genes)[as.character(strand(genes)) == "-"] - 1
  start(downstream)[as.character(strand(genes)) == "-"] <-
    start(genes)[as.character(strand(genes)) == "-"] - distance
  
  return(downstream)
}

downstream_regions <- gtf |>
  genes() |>
  get_downstream_regions(3000) |>
  GenomicRanges::reduce() |>
  GenomicRanges::setdiff(c(
    promoter_regions, three_prime_utr_regions, exonic_regions, intronic_regions
  ))

downstream_sum <- downstream_regions |>
  width() |>
  sum()

# Caluclate the distal intergenic length.
distal_intergenic_length <-
  genome_length - (
    promoter_sum +
    three_prime_utr_sums +
    exonic_sum +
    intronic_sum +
    downstream_sum    
  )

# Get the percentage of the genome that is each feature.
feature_perc <-
  tibble(
    sample_name="Mouse Genome",
    annotation=c(
      "Promoter", "3' UTR", "Exon", "Intron", "Downstream", "Distal Intergenic"
    ),
    n=c(
      promoter_sum,
      three_prime_utr_sums,
      exonic_sum,
      intronic_sum,
      downstream_sum,
      distal_intergenic_length
    )
  ) |>
  mutate(frac=n/sum(n))
```

Finally we will plot the distribution of insertion sites relative to genomic
features for each sample. The overall distribution of genomic features
calculated earlier will will be included as reference.

```{r, fig.height=4.5, fig.width=5, warning=FALSE}
annotation_factor_order <- c(
  "Promoter", "Exon", "Intron", "3' UTR", "Downstream", "Distal Intergenic"
)

name_order <- c(
  "Mouse BM (65.10%)", "Mouse BM (62.20%)", "Mouse BM (38.90%)",
  "Mouse BM (37.00%)", "Mouse BM (25.00%)", "Mouse BM (0.00%)",
  "Mouse Genome"
)

p <- isa |>
  # Remove superfluous parts of the annotation values.
  mutate(annotation=str_remove(annotation, "\\s\\(.+$")) |>
  # Get the fraction of insertion sites associated with each genomic annotation.
  count(sample_name, annotation) |>
  group_by(sample_name) |>
  mutate(frac=n/sum(n)) |>
  ungroup() |>
  complete(sample_name, annotation, fill=list(n=0)) |>
  # Add the general mouse genome stats.
  bind_rows(feature_perc) |>
  mutate(
    annotation=factor(annotation, levels=annotation_factor_order),
    sample_name=factor(sample_name, levels=name_order)
  ) |>
  # Create the stacked barplot.
  ggplot(aes(x=frac, y=fct_rev(sample_name))) +
    geom_col(aes(fill=fct_rev(annotation))) +
    scale_fill_brewer(palette="Set1") +
    scale_x_continuous(labels=scales::percent) +
    theme_bw() +
    theme(
      legend.position="bottom"
    ) +
    guides(fill=guide_legend(ncol=2)) +
    labs(
      title="Genomic Distribution of\nInsertion Sites",
      x="Percentage of Insertion Site\nLocations",
      y="Tissue (GFP+ %)",
      fill="Genomic\nAnnotation"
    )

print(p)

ggsave(
  file.path("results", "insertion_site_genomic_locations.pdf"),
  plot=p, device=cairo_pdf, height=4.5, width=5
)
```

For all insertion sites appearing in at least two technical replicates,
create a sequence logo centered around the insertion site including ±20
bps.

Sleeping beauty is known to have a palindromic motif at the insertion
site with A(-3), T(0), A(+1), and T(+4).

```{r, fig.height=8, fig.width=8, warning=FALSE, message=FALSE}
# Load the FASTA data into a table.
fasta_files <- 
  file.path("data", "insertion_site_sequences") |>
  list.files(pattern="\\.fasta$", full.names=TRUE) |>
  tibble(file_path=_) |>
  mutate(
    file_id=str_remove(basename(file_path), "_S\\d+_L00\\d\\.\\S+"),
    data=map(file_path, readDNAStringSet),
    posid=map(data, names),
    seq=map(data, \(x) {unname(as.character(x))})
  ) |>
  left_join(
    dplyr::select(sample_sheet, file_id, sample_name),
    by="file_id"
  ) |>
  dplyr::select(!c(data, file_path, file_id)) |>
  unnest(c(posid, seq)) |>
  distinct() |>
  separate_wider_regex(
    posid, c(
      seqnames="^\\S+", ":", start="\\d+", "-", end="\\d+", "\\(",
      strand="[\\+\\-]", "\\)"
    ),
    cols_remove=FALSE
  ) |>
  mutate(across(c(start, end), as.numeric))

# Remove insertion sites that appeared in less than 2 technical replicates.
fasta_files <- semi_join(
  fasta_files,
  mutate(isa, start=start - 21, end=end + 20),
  by=c("sample_name", "seqnames", "start", "end", "strand")
)

# Turn it into a named list for ggseqlogo.
fasta_files <- fasta_files |>
  split(~sample_name) |>
  map(pull, seq)

# Order the sequences.
fasta_files <- fasta_files[levels(isa$sample_name)]

# Create the sequence logos.
p <- ggseqlogo(fasta_files, ncol=1, seq_type="dna") +
  theme(
    text=element_text(size=7)
  ) +
  scale_x_continuous(breaks=seq_len(41), labels=seq(-20, 20, 1)) +
  labs(
    x="Position Relative to\nInsertion Site"
  )

print(p)

ggsave(
  file.path("results", "insertion_site_seqlogo.pdf"),
  plot=p, device=cairo_pdf, height=8, width=8
)
```

For each sample plot the location of the insertion sites relative to
their chromosomal location on a karyogram.

```{r, fig.height=5, fig.width=6}
# Create GRanges for each sample.
is_granges <- isa |>
  dplyr::select(sample_name, seqnames, start, end, strand, score) |>
  mutate(seqnames=str_c("chr", seqnames)) |>
  split(~ sample_name) |>
  map(makeGRangesFromDataFrame, keep.extra.columns=TRUE)

if (!dir.exists(file.path("results", "karyoplots"))) {
  dir.create(file.path("results", "karyoplots"))
}

# Increase the top plot margin so the title doesn't get cut off.
pp <- getDefaultPlotParams(plot.type=1)
pp$topmargin <- 300

# Create a plot for each sample.
iwalk(is_granges, \(x, y) {
  species <- ifelse(str_starts(y, "Human"), "hg38", "mm39")

  file_name <- y |>
    str_to_lower() |>
    str_replace_all(c(" "="_", "%"="")) |>
    str_glue("{file}_karyoplot.pdf", file=_)

  par(mar=c(1, 1, 10, 1))
  cairo_pdf(
    filename=file.path("results", "karyoplots", file_name),
    height=5, width=6
  )
  
  p <- plotKaryotype(genome=species, main=y, plot.params=pp)
  kpSegments(p, x, y1=0.5, y0=0, col="royalblue")
  
  dev.off()
})
```

## Software Information

```{r, collapse=FALSE, echo=FALSE, comment=""}
sessionInfo()
```
